File: retort-core/src/main/java/co/kaioru/retort/util/builder/CommandBuilder.java
package co.kaioru.retort.util.builder;

import co.kaioru.retort.command.ICommand;
import co.kaioru.retort.command.ICommandExecutable;
import com.google.common.collect.Lists;

import java.util.List;

public abstract class CommandBuilder<T extends CommandBuilder, C extends ICommand, E extends ICommandExecutable> {

	protected final String name;
	protected String desc;
	protected List<String> aliases;
	protected List<ICommand> commands;

	public CommandBuilder(String name) {
		this.name = name;
		this.desc = "No description";
		this.aliases = Lists.newArrayList();
		this.commands = Lists.newArrayList();
	}

	public T desc(String desc) {
		this.desc = desc;
		return (T) this;
	}

	public T alias(String alias) {
		this.aliases.add(alias);
		return (T) this;
	}

	public T command(ICommand command) {
		this.commands.add(command);
		return (T) this;
	}

	public abstract C build(E executable);

}

File: retort-core/src/main/kotlin/co/kaioru/retort/builder/CommandBuilder.kt
package co.kaioru.retort.builder

import co.kaioru.retort.Command
import co.kaioru.retort.CommandContext
import co.kaioru.retort.ICommand
import co.kaioru.retort.ICommandExecutable

class CommandBuilder<I : CommandContext, O>(name: String) : Command<I, O>(name), ICommandBuilder<I, O> {
    var executable: ICommandExecutable<I, O>? = null

    override fun build(executable: ICommandExecutable<I, O>): ICommand<I, O> {
        this.executable = executable
        return this
    }

    override fun execute(input: I): O {
        if (executable == null)
            TODO()
        return executable?.execute(input)!!
    }
}